File: src/main/java/nodebox/graphics/Color.java
/*
 * This file is part of NodeBox.
 *
 * Copyright (C) 2008 Frederik De Bleser (frederik@pandora.be)
 *
 * NodeBox is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * NodeBox is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with NodeBox. If not, see <http://www.gnu.org/licenses/>.
 */

package nodebox.graphics;

import java.awt.*;

import static com.google.common.base.Preconditions.checkArgument;
import static nodebox.graphics.MathUtils.clamp;

public final class Color implements Cloneable {

    public enum Mode {
        RGB, HSB, CMYK
    }

    public static final Color BLACK = new Color(0);
    public static final Color WHITE = new Color(1);

    private final double r, g, b, a;
    private final double h, s, v;
    private transient java.awt.Color awtColor = null;

    public static Color fromHSB(double hue, double saturation, double brightness) {
        return new Color(hue, saturation, brightness, Mode.HSB);
    }

    public static Color fromHSB(double hue, double saturation, double brightness, double alpha) {
        return new Color(hue, saturation, brightness, alpha, Mode.HSB);
    }

    public static Color valueOf(String hex) {
        return new Color(hex);
    }

    /**
     * Create an empty (black) color object.
     */
    public Color() {
        this(0, 0, 0, 1.0, Mode.RGB);
    }

    /**
     * Create a new color with the given grayscale value.
     *
     * @param v the gray component.
     */
    public Color(double v) {
        this(v, v, v, 1.0, Mode.RGB);
    }

    /**
     * Create a new color with the given grayscale and alpha value.
     *
     * @param v the grayscale value.
     * @param a the alpha value.
     */

    public Color(double v, double a) {
        this(v, v, v, a, Mode.RGB);
    }

    /**
     * Create a new color with the the given R/G/B value.
     *
     * @param x the red or hue component.
     * @param y the green or saturation component.
     * @param z the blue or brightness component.
     * @param m the specified color mode.
     */
    public Color(double x, double y, double z, Mode m) {
        this(x, y, z, 1.0, m);
    }

    /**
     * Create a new color with the the given R/G/B value.
     *
     * @param r the red component.
     * @param g the green component.
     * @param b the blue component.
     */
    public Color(double r, double g, double b) {
        this(r, g, b, 1.0, Mode.RGB);
    }

    /**
     * Create a new color with the the given R/G/B/A or H/S/B/A value.
     *
     * @param r the red component.
     * @param g the green component.
     * @param b the blue component.
     * @param a the alpha component.
     */
    public Color(double r, double g, double b, double a) {
        this(r, g, b, a, Mode.RGB);
    }

    /**
     * Create a new color with the the given R/G/B/A or H/S/B/A value.
     *
     * @param x the red or hue component.
     * @param y the green or saturation component.
     * @param z the blue or brightness component.
     * @param a the alpha component.
     * @param m the specified color mode.
     */
    public Color(double x, double y, double z, double a, Mode m) {
        checkArgument(m != Mode.CMYK, "CMYK not supported");

        switch (m) {
            case CMYK:
                throw new RuntimeException("CMYK color mode is not implemented yet.");
            case HSB:
                this.h = clamp(x);
                this.s = clamp(y);
                this.v = clamp(z);
                this.a = clamp(a);
                double[] rgb = updateRGB();
                this.r = rgb[0];
                this.g = rgb[1];
                this.b = rgb[2];
                //updateCMYK();
                break;
            case RGB:
            default:
                this.r = clamp(x);
                this.g = clamp(y);
                this.b = clamp(z);
                this.a = clamp(a);
                double[] hsb = updateHSB();
                this.h = hsb[0];
                this.s = hsb[1];
                this.v = hsb[2];
                //updateCMYK();
                break;
        }
    }

    public Color(String colorName) {
        if (!colorName.startsWith("#")) {
            throw new IllegalArgumentException("The given value '" + colorName + "' is not of the format #112233.");
        }
        int r255, g255, b255, a255 = 255;
        if (colorName.length() == 4) { // #123
            r255 = Integer.parseInt(colorName.substring(1, 2) + colorName.substring(1, 2), 16);
            g255 = Integer.parseInt(colorName.substring(2, 3) + colorName.substring(2, 3), 16);
            b255 = Integer.parseInt(colorName.substring(3, 4) + colorName.substring(3, 4), 16);
        } else if (colorName.length() == 5) { // #123f
            r255 = Integer.parseInt(colorName.substring(1, 2) + colorName.substring(1, 2), 16);
            g255 = Integer.parseInt(colorName.substring(2, 3) + colorName.substring(2, 3), 16);
            b255 = Integer.parseInt(colorName.substring(3, 4) + colorName.substring(3, 4), 16);
            a255 = Integer.parseInt(colorName.substring(4, 5) + colorName.substring(4, 5), 16);
        } else if (colorName.length() == 7) { // #112233
            r255 = Integer.parseInt(colorName.substring(1, 3), 16);
            g255 = Integer.parseInt(colorName.substring(3, 5), 16);
            b255 = Integer.parseInt(colorName.substring(5, 7), 16);
        } else if (colorName.length() == 9) { // #112233ff
            r255 = Integer.parseInt(colorName.substring(1, 3), 16);
            g255 = Integer.parseInt(colorName.substring(3, 5), 16);
            b255 = Integer.parseInt(colorName.substring(5, 7), 16);
            a255 = Integer.parseInt(colorName.substring(7, 9), 16);
        } else {
            throw new IllegalArgumentException("The given value '" + colorName + "' is not of the format #112233.");
        }
        this.r = r255 / 255.0;
        this.g = g255 / 255.0;
        this.b = b255 / 255.0;
        this.a = a255 / 255.0;
        double[] hsb = updateHSB();
        this.h = hsb[0];
        this.s = hsb[1];
        this.v = hsb[2];
        //updateCMYK();
    }

    /**
     * Create a new color with the the given color.
     * <p/>
     * The color object is cloned; you can change the original afterwards.
     * If the color object is null, the new color is turned off (same as nocolor).
     *
     * @param color the color object.
     */
    public Color(java.awt.Color color) {
        this(color.getRed() / 255.0, color.getGreen() / 255.0, color.getBlue() / 255.0, color.getAlpha() / 255.0);
    }

    /**
     * Create a new color with the the given color.
     * <p/>
     * The color object is cloned; you can change the original afterwards.
     * If the color object is null, the new color is turned off (same as nocolor).
     *
     * @param other the color object.
     */
    public Color(Color other) {
        this(other.r, other.g, other.b, other.a);
    }

    public double getRed() {
        return r;
    }

    public double getR() {
        return r;
    }

    public double getGreen() {
        return g;
    }

    public double getG() {
        return g;
    }

    public double getBlue() {
        return b;
    }

    public double getB() {
        return b;
    }

    public double getAlpha() {
        return a;
    }

    public double getA() {
        return a;
    }

    public boolean isVisible() {
        return a > 0.0;
    }

    public double getHue() {
        return h;
    }

    public double getH() {
        return h;
    }

    public double getSaturation() {
        return s;
    }

    public double getS() {
        return s;
    }

    public double getBrightness() {
        return v;
    }

    public double getV() {
        return v;
    }

    private double[] updateRGB() {
        if (s == 0)
            return new double[]{this.v, this.v, this.v};
        else {
            double h = this.h;
            if (this.h == 1.0)
                h = 0.999998;
            double s = this.s;
            double v = this.v;
            double f, p, q, t;
            h = h / (60.0 / 360);
            int i = (int) Math.floor(h);
            f = h - i;
            p = v * (1 - s);
            q = v * (1 - s * f);
            t = v * (1 - s * (1 - f));

            double rgb[];
            if (i == 0)
                rgb = new double[]{v, t, p};
            else if (i == 1)
                rgb = new double[]{q, v, p};
            else if (i == 2)
                rgb = new double[]{p, v, t};
            else if (i == 3)
                rgb = new double[]{p, q, v};
            else if (i == 4)
                rgb = new double[]{t, p, v};
            else
                rgb = new double[]{v, p, q};

            return new double[]{rgb[0], rgb[1], rgb[2]};
        }
    }

    private double[] updateHSB() {
        double h = 0;
        double s = 0;
        double v = Math.max(Math.max(r, g), b);
        double d = v - Math.min(Math.min(r, g), b);

        if (v != 0)
            s = d / v;

        if (s != 0) {
            if (r == v)
                h = 0 + (g - b) / d;
            else if (g == v)
                h = 2 + (b - r) / d;
            else
                h = 4 + (r - g) / d;
        }

        h = h * (60.0 / 360);
        if (h < 0)
            h = h + 1;

        return new double[]{h, s, v};
    }

    private void updateCMYK() {
        // TODO: implement
    }

    public java.awt.Color getAwtColor() {
        // We don't return the cached awtColor here, since java.awt.Color is mutable.
        return new java.awt.Color((float) getRed(), (float) getGreen(), (float) getBlue(), (float) getAlpha());
    }

    public void set(Graphics2D g) {
        if (awtColor == null) {
            awtColor = new java.awt.Color((float) getRed(), (float) getGreen(), (float) getBlue(), (float) getAlpha());
        }
        g.setColor(awtColor);
    }

    @Override
    public Color clone() {
        return new Color(this);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Color)) return false;
        Color other = (Color) obj;
        // Because of the conversion to/from hex, we can have rounding errors.
        // Therefore, we only compare what we can store, i.e. values in the 0-255 range.
        return Math.round(r * 255) == Math.round(other.r * 255)
                && Math.round(g * 255) == Math.round(other.g * 255)
                && Math.round(b * 255) == Math.round(other.b * 255)
                && Math.round(a * 255) == Math.round(other.a * 255);
    }

    /**
     * Parse a hexadecimal value and return a Color object.
     * <p/>
     * The value needs to have four components. (R,G,B,A)
     *
     * @param value the hexadecimal color value, e.g. #995423ff
     * @return a Color object.
     */
    public static Color parseColor(String value) {
        return new Color(value);
    }

    private String paddedHexString(int v) {
        String s = Integer.toHexString(v);
        if (s.length() == 1) {
            return "0" + s;
        } else if (s.length() == 2) {
            return s;
        } else {
            throw new AssertionError("Value too large (must be between 0-255, was " + v + ").");
        }
    }

    /**
     * Returns the color as a 8-bit hexadecimal value, e.g. #ae45cdff
     *
     * @return the color as a 8-bit hexadecimal value
     */
    @Override
    public String toString() {
        int r256 = (int) Math.round(r * 255);
        int g256 = (int) Math.round(g * 255);
        int b256 = (int) Math.round(b * 255);
        int a256 = (int) Math.round(a * 255);
        return "#"
                + paddedHexString(r256)
                + paddedHexString(g256)
                + paddedHexString(b256)
                + paddedHexString(a256);
    }

    public String toCSS() {
        if (!isVisible()) {
            return "none";
        }
        StringBuilder sb = new StringBuilder();
        int r256 = (int) Math.round(r * 255);
        int g256 = (int) Math.round(g * 255);
        int b256 = (int) Math.round(b * 255);
        if (a == 1.0) {
            return "#"
                    + paddedHexString(r256)
                    + paddedHexString(g256)
                    + paddedHexString(b256);
        } else {
            return String.format("rgba(%d,%d,%d,%.2f)", r256, g256, b256, a);
        }
    }
}

File: packages/core/src/graphics/color.ts
import { clamp } from "./math";

/**
 * An RGBA color with components in 0..1, plus derived HSB values (nodebox.graphics.Color).
 * Immutable. Serializes to "#rrggbbaa", the .ndbx format.
 */
export class Color {
  static readonly BLACK = new Color(0, 0, 0, 1);
  static readonly WHITE = new Color(1, 1, 1, 1);

  readonly r: number;
  readonly g: number;
  readonly b: number;
  readonly a: number;
  readonly h: number;
  readonly s: number;
  readonly v: number;

  constructor(r = 0, g = 0, b = 0, a = 1) {
    this.r = clamp(r, 0, 1);
    this.g = clamp(g, 0, 1);
    this.b = clamp(b, 0, 1);
    this.a = clamp(a, 0, 1);
    const [h, s, v] = rgbToHsb(this.r, this.g, this.b);
    this.h = h;
    this.s = s;
    this.v = v;
  }

  static gray(v: number, a = 1): Color {
    return new Color(v, v, v, a);
  }

  static fromHSB(hue: number, saturation: number, brightness: number, alpha = 1): Color {
    const [r, g, b] = hsbToRgb(clamp(hue, 0, 1), clamp(saturation, 0, 1), clamp(brightness, 0, 1));
    return new Color(r, g, b, alpha);
  }

  /** Parse #rgb, #rgba, #rrggbb or #rrggbbaa. */
  static parse(value: string): Color {
    if (!value.startsWith("#")) throw new Error(`The given value '${value}' is not of the format #112233.`);
    const hex = value.slice(1);
    let r: number, g: number, b: number, a = 255;
    if (hex.length === 3 || hex.length === 4) {
      r = parseInt(hex[0] + hex[0], 16);
      g = parseInt(hex[1] + hex[1], 16);
      b = parseInt(hex[2] + hex[2], 16);
      if (hex.length === 4) a = parseInt(hex[3] + hex[3], 16);
    } else if (hex.length === 6 || hex.length === 8) {
      r = parseInt(hex.slice(0, 2), 16);
      g = parseInt(hex.slice(2, 4), 16);
      b = parseInt(hex.slice(4, 6), 16);
      if (hex.length === 8) a = parseInt(hex.slice(6, 8), 16);
    } else {
      throw new Error(`The given value '${value}' is not of the format #112233.`);
    }
    if ([r, g, b, a].some((c) => Number.isNaN(c))) {
      throw new Error(`The given value '${value}' is not of the format #112233.`);
    }
    return new Color(r / 255, g / 255, b / 255, a / 255);
  }

  static valueOf(value: string): Color {
    return Color.parse(value);
  }

  static isColor(value: unknown): value is Color {
    return value instanceof Color;
  }

  get red(): number {
    return this.r;
  }
  get green(): number {
    return this.g;
  }
  get blue(): number {
    return this.b;
  }
  get alpha(): number {
    return this.a;
  }
  get hue(): number {
    return this.h;
  }
  get saturation(): number {
    return this.s;
  }
  get brightness(): number {
    return this.v;
  }

  isVisible(): boolean {
    return this.a > 0;
  }

  withAlpha(a: number): Color {
    return new Color(this.r, this.g, this.b, a);
  }

  equals(other: unknown): boolean {
    return other instanceof Color && other.r === this.r && other.g === this.g && other.b === this.b && other.a === this.a;
  }

  /** "#rrggbbaa", the .ndbx serialization. */
  toString(): string {
    return "#" + hex2(this.r) + hex2(this.g) + hex2(this.b) + hex2(this.a);
  }

  toHex(): string {
    return this.toString();
  }

  toCSS(): string {
    if (!this.isVisible()) return "none";
    if (this.a === 1) return "#" + hex2(this.r) + hex2(this.g) + hex2(this.b);
    return `rgba(${Math.round(this.r * 255)},${Math.round(this.g * 255)},${Math.round(this.b * 255)},${this.a.toFixed(2)})`;
  }

  toJSON(): { r: number; g: number; b: number; a: number } {
    return { r: this.r, g: this.g, b: this.b, a: this.a };
  }
}

function hex2(v: number): string {
  return Math.round(v * 255)
    .toString(16)
    .padStart(2, "0");
}

function hsbToRgb(h: number, s: number, v: number): [number, number, number] {
  if (s === 0) return [v, v, v];
  if (h === 1.0) h = 0.999998;
  h = h / (60.0 / 360);
  const i = Math.floor(h);
  const f = h - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));
  switch (i) {
    case 0:
      return [v, t, p];
    case 1:
      return [q, v, p];
    case 2:
      return [p, v, t];
    case 3:
      return [p, q, v];
    case 4:
      return [t, p, v];
    default:
      return [v, p, q];
  }
}

function rgbToHsb(r: number, g: number, b: number): [number, number, number] {
  let h = 0;
  let s = 0;
  const v = Math.max(r, g, b);
  const d = v - Math.min(r, g, b);
  if (v !== 0) s = d / v;
  if (s !== 0) {
    if (r === v) h = 0 + (g - b) / d;
    else if (g === v) h = 2 + (b - r) / d;
    else h = 4 + (r - g) / d;
  }
  h = h * (60.0 / 360);
  if (h < 0) h = h + 1;
  return [h, s, v];
}
